### PR TITLE
WIP: Add snirf writer

### PR DESCRIPTION
**Pull request by rob-luke**
Add support for writing SNIRF files.

https://github.com/mne-tools/mne-python/issues/8060

### CODE REVIEW

**Comment by rob-luke** [comment]
Hi @larsoner, would you be able to give this a first review? As with the reader, we are only covering a small part of the SNIRF format. But at a minimum I want to be able to read a NIRX file, write it as SNIRF, and read back without loss. Thanks

I think the CI failure is unrelated and due to some other change in MNE plot_alignment at master.

File: mne_nirs/io/__init__.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+"""File IO."""
+
+# Authors: Robert Luke <mail@robertluke.net>
+#
+# License: BSD (3-clause)
+
+from . import snirf
```
Comment (agramfort) [inline mne_nirs/io/__init__.py:7]: missing empty line

File: mne_nirs/io/snirf/__init__.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+"""SNIRF module for conversion to FIF."""
+
+# Author: Robert Luke <mail@robertluke.net>
+#
+# License: BSD (3-clause)
+
+from ._snirf import write_raw_snirf
```
Comment (agramfort) [inline mne_nirs/io/snirf/__init__.py:7]: idem

File: mne_nirs/io/snirf/_snirf.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
+import h5py as h5py
```
Comment (agramfort) [inline mne_nirs/io/snirf/_snirf.py:1]: authors and licence missing

File: mne_nirs/io/snirf/_snirf.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    with h5py.File(fname, "w") as f:
+        f.create_dataset("nirs/data1/measurementList1/dataType", data=1)
+        f.create_dataset("/nirs/data1/dataTimeSeries", data=raw.get_data().T)
+        f.create_dataset("/nirs/data1/time", data=raw.times)
+
+        # Extract info from file names
+        rgx = r'S(\d+)_D(\d+) (\d+)'
+        chs = raw.info['chs']
+        sources = [float(mch(rgx, r['ch_name']).groups()[0])for r in chs]
```
Comment (agramfort) [inline mne_nirs/io/snirf/_snirf.py:34]:
```suggestion
        sources = [float(mch(rgx, r['ch_name']).groups()[0]) for r in chs]
```

File: mne_nirs/io/snirf/_snirf.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        f.create_dataset("nirs/probe/detectorPos3D", data=detlocs)
+        f.create_dataset("nirs/metaDataTags/LengthUnit",
+                         data=['m'.encode('UTF-8')])
+
+        # Prep data for storing each MNE channel as SNIRF measurementList
+        channels = ["measurementList" + str(idx + 1)
+                    for idx in range(len(raw.ch_names))]
+        sources = np.array([float(src) for src in sources])
+        detectors = np.array([float(det) for det in detectors])
+        sources_sorted = list([float(src) for src in sources_sorted])
```
Comment (agramfort) [inline mne_nirs/io/snirf/_snirf.py:79]:
```suggestion
        sources_sorted = [float(src) for src in sources_sorted]
```

File: mne_nirs/io/snirf/_snirf.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        detectors_sorted = list([float(det) for det in detectors_sorted])
```
Comment (agramfort) [inline mne_nirs/io/snirf/_snirf.py:80]: idem

File: mne_nirs/io/snirf/_snirf.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        for idx, ch in enumerate(channels):
+            f.create_dataset('nirs/data1/' + ch + '/sourceIndex',
+                             data=[sources_sorted.index(sources[idx]) + 1])
+            f.create_dataset('nirs/data1/' + ch + '/detectorIndex',
+                             data=[detectors_sorted.index(detectors[idx]) + 1])
+            f.create_dataset('nirs/data1/' + ch + '/wavelengthIndex',
+                             data=[wavelengths_index[idx]])
+
+        # Store demographic info
+        SubjectID = raw.info["subject_info"]['first_name']
```
Comment (agramfort) [inline mne_nirs/io/snirf/_snirf.py:94]:
```suggestion
        subject_id = raw.info["subject_info"]['first_name']
```

File: mne_nirs/io/snirf/tests/test_snirf.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+                         'NIRx', 'nirx_15_2_recording_w_short')
+
+
+@requires_testing_data
+@requires_h5py
+def test_snirf_write(tmpdir):
+    """Test reading NIRX files."""
+    raw_orig = read_raw_nirx(fname_original, preload=True)
+    write_raw_snirf(raw_orig, tmpdir.join('test_raw.snirf'))
+    raw = read_raw_snirf(tmpdir.join('test_raw.snirf'))
```
Comment (agramfort) [inline mne_nirs/io/snirf/tests/test_snirf.py:28]: I would just use object_diff on the info before and after IO round trip.

**Comment by rob-luke** [comment]
Thanks @agramfort

**Comment by rob-luke** [comment]
Thanks for the tip @agramfort about `object_diff` it has helped me to identify the following difference between the original and what I read back...

```python
['chs'][1]['logno'] value mismatch (14, 2)
['chs'][1]['scanno'] value mismatch (14, 2)
['chs'][2]['logno'] value mismatch (2, 3)
['chs'][2]['scanno'] value mismatch (2, 3)
['chs'][3]['logno'] value mismatch (15, 4)
...
```

I can't find much information about scanno and logno other than what's here https://mne.tools/stable/generated/mne.Info.html?highlight=scanno 
 
Is it worth digging in to making these the same between the original and the snirf writen/read?

File: mne_nirs/io/snirf/_snirf.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            f.create_dataset('nirs/data1/' + ch + '/sourceIndex',
+                             data=[sources_sorted.index(sources[idx]) + 1])
+            f.create_dataset('nirs/data1/' + ch + '/detectorIndex',
+                             data=[detectors_sorted.index(detectors[idx]) + 1])
+            f.create_dataset('nirs/data1/' + ch + '/wavelengthIndex',
+                             data=[wavelengths_index[idx]])
+
+        # Store demographic info
+        subject_id = raw.info["subject_info"]['first_name']
+        f.create_dataset("nirs/metaDataTags/SubjectID",
```
Comment (rob-luke) [inline mne_nirs/io/snirf/_snirf.py:99]: SNIRF has a `SubjectID` field. Currently I store the `first_name` here. But then we loose `middle_name` and `last_name` in the round trip. Other options are:

1. Save `first_name` + ` ` + `middle_name` + ` ` + `last_name` in this field then modify the SNIRF reader to split by spaces.
2. Add our own metadata tags as the specification states `We do not limit the total number of metadata records in the metaDataTags. Users can add additional customized metadata records`.

I am leaning towards option 2

**Comment by agramfort** [comment]
@rob-luke you should not open PR from the upstream repo. Use your fork

**Comment by rob-luke** [comment]
> @rob-luke you should not open PR from the upstream repo. Use your fork

Sorry. Will do from now on.

**Pull request closed without merging**